### PR TITLE
ARROW-7346: [CI] Explicit usage of ccache across the builds

### PR DESCRIPTION
**Pull request by kszucs**
- ccache configuration is centralised in the docker-compose.yml
- stricter build directory paths
- java jni tests were not running, since the vendoring of flatbuffers there was a hidden linking error

### CODE REVIEW

File: ci/docker/cuda-10.0-cpp.dockerfile [hunk 1]
```diff
@@ -72,7 +72,9 @@ RUN apt-get update -y -q && \
 # - libgtest-dev only provide sources
 # - libprotobuf-dev only provide sources
 # - thrift is too old
-ENV ARROW_DEPENDENCY_SOURCE=SYSTEM \
+ENV PATH=/usr/lib/ccache/:$PATH \
```
Comment (kou) [inline ci/docker/cuda-10.0-cpp.dockerfile]: Could you sort this list alphabetically?

**Review by kou — APPROVED** [review]
+1

File: ci/docker/cuda-10.0-cpp.dockerfile (hunk 1 above)
Comment (kszucs) [reply ci/docker/cuda-10.0-cpp.dockerfile]: Sure

**Review by kszucs — COMMENTED** [review]
+1

**Pull request closed without merging**